content/lessons/2_ownership/index.md

```diff
@@ -0,0 +1,346 @@
+
+```sh
+error[E0382]: borrow of moved value: `s`
+ --> src/main.rs:6:42
+  |
+2 |     let s = String::from("a string");
+  |         - move occurs because `s` has type `String`, which does not implement the `Copy` trait
+3 |
+4 |     let s2 = s;
+  |              - value moved here
+5 |
+6 |     println!("And the contents are: {}", s);
+  |                                          ^ value borrowed here after move
+```
+
+_"value moved here"_ - gotcha! So `s` is being moved to `s2`, which also means that `s2` now becomes the new owner of the string being moved and `s` is left in an invalid state. In Rust, the default method of passing values around is by move, not by copy. While it may sound a bit odd at first, it actually has some very interesting implications. But before we get to them, let's fix our code so it compiles now. To do so, we have to explicitly tell Rust to make a copy by invoking the `clone` method:
+
+```rust
+let s = String::from("a string"); // 's' is an owner
+
+let s2 = s.clone(); // 's2' now contains its own copy
+
+println!("And the contents are: {}", s); // success!
+```
+
+The compiler is happy now and so are we. The implicit move takes some getting used to, but the compiler is here to help us. Now, let's put the good, old C++ on the table again and compare the two lines:
+
+<div style="text-align: center">
+
+`let s2 = s;` is equivalent to `auto s2 = std::move(s);`
+
+`let s2 = s.clone()` is equivalent to `auto s2 = s`
+
+</div>
+
+There are a few important things to note here:
+
+- Making a copy is oftentimes not cheap. Memory needs to be allocated and copied, and a call to the system has to be made. We should prefer to move things as much as possible to avoid this cost - in C++ we have a myriad of language features like `std::move` and _r-references_ to achieve this. Every programmer worth their salt needs to be well versed in all of them to write efficient C++ code and simply forgetting one move can lead to significant performance loss (and this happens to even the most senior devs ever existing, let's not pretend). On the contrary, in Rust you need to make an effort to make a copy and that makes you very aware of the cost you're paying - something that we'll see quite a lot of in the language. Also, if you forget a clone there's no harm done - it just won't compile!
+
+- Hidden in all of this is another nice thing Rust gives us. In C++, nothing prevents you from using variables after they've been moved from, leading to unexpected errors in a more complex code. In Rust, that variable (in our case `s`) simply becomes invalid and the compiler gives us a nice error about it.
+
+### But what about ints?
+
+A good question to ask. Copying primitives is cheap, cheaper even than moving them. It's not convenient for the programmer either to have to always write `.clone()` after every primitive. If we take a look at the error from the previous example:
```

Comment:
> Copying primitives is cheap, cheaper even than moving them.

It's not true, in Rust moving and copying are both implemented as a simple memcpy, the difference is only in the restraints imposed by the compiler